src/wp-includes/html-api/class-wp-html-processor.php

```diff
@@ -424,6 +424,116 @@ function ( WP_HTML_Token $token ): void {
 		};
 	}
 
+	/**
+	 * Creates a fragment processor at the current node.
+	 *
+	 * HTML Fragment parsing always happens with a context node. HTML Fragment Processors can be
+	 * instantiated with a `BODY` context node via `WP_HTML_Processor::create_fragment( $html )`.
+	 *
+	 * The context node may impact how a fragment of HTML is parsed. For example, consider the HTML
+	 * fragment `<td />Inside TD?</td>`.
+	 *
+	 * A BODY context node will produce the following tree:
+	 *
+	 *     └─#text Inside TD?
+	 *
+	 * Notice that the `<td>` tags are completely ignored.
+	 *
+	 * Compare that with an SVG context node that produces the following tree:
+	 *
+	 *     ├─svg:td
+	 *     └─#text Inside TD?
+	 *
+	 * Here, a `td` node in the `svg` namespace is created, and its self-closing flag is respected.
+	 * This is a peculiarity of parsing HTML in foreign content like SVG.
+	 *
+	 * Finally, consider the tree produced with a TABLE context node:
+	 *
+	 *     └─TBODY
+	 *       └─TR
+	 *         └─TD
+	 *           └─#text Inside TD?
+	 *
+	 * These examples demonstrate how important the context node may be when processing an HTML
+	 * fragment. Special care must be taken when processing fragments that are expected to appear
+	 * in specific contexts. SVG and TABLE are good examples, but there are others.
+	 *
+	 * @see https://html.spec.whatwg.org/multipage/parsing.html#html-fragment-parsing-algorithm
+	 *
+	 * @param string $html Input HTML fragment to process.
+	 * @return static|null The created processor if successful, otherwise null.
+	 */
+	public function create_fragment_at_current_node( string $html ) {
+		if ( $this->get_token_type() !== '#tag' ) {
+			return null;
+		}
+
+		$namespace = $this->current_element->token->namespace;
+
+		/*
+		 * Prevent creating fragments at nodes that require a special tokenizer state.
+		 * This is unsupported by the HTML Processor.
+		 */
+		if (
+			'html' === $namespace &&
+			in_array( $this->current_element->token->node_name, array( 'IFRAME', 'NOEMBED', 'NOFRAMES', 'SCRIPT', 'STYLE', 'TEXTAREA', 'TITLE', 'XMP', 'PLAINTEXT' ), true )
+		) {
+			return null;
+		}
+
+		$fragment_processor = static::create_fragment( $html );
+		if ( null === $fragment_processor ) {
+			return null;
+		}
+
+		$fragment_processor->compat_mode = $this->compat_mode;
+
+		$fragment_processor->context_node                = clone $this->state->current_token;
+		$fragment_processor->context_node->bookmark_name = 'context-node';
+		$fragment_processor->context_node->on_destroy    = null;
+
+		$fragment_processor->state->context_node = array( $fragment_processor->context_node->node_name, array() );
+		foreach ( $this->get_attribute_names_with_prefix( '' ) as $name => $value ) {
```

Comment:
`get_attribute_names_with_prefix` returns an array listing attribute names or null. What's the reason for iterating over this list instead of assigning it as is a line before?

Reply:
Great spot, thank you! I had confused this with the iterator-based methods like `class_list`. I think you've actually spotted a bug here I'll see about adding a test and fix for it.

Reply:
I was about to ask whether any of the examples included in https://github.com/WordPress/wordpress-develop/pull/7348#issuecomment-2474165485 could be used to test this behavior. Some of them might be covered by the html5lib test suite though. I'm also happy to write some tests as follow-ups if you can help clarify what makes the most sense.

Reply:
I suspect the reason this didn't impact anything is that we don't use `state->context_node`. I think it's been superseded by the `context_node` property on the processor and is redundant.

I did fix this behavior but I don't think it can really be well tested without just looking at internals.

I'll propose removing the property in a future PR.

Reply:
The fix looks good and now the loop is fully justified 👍